### PR TITLE
Initial commit of data preprocessing code

### PR DESCRIPTION
**Pull request by najdamikolaj00**
The entire package for signal preprocessing used in AUTOMATIC project.

### CODE REVIEW

File: Code/utilities.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            print("Creating a new JSON file or overwriting malformed content.")
+        
+        existing_data[signal_name] = data
+        
+        with open(file_path, 'w') as file:
+            json.dump(existing_data, file, indent=4)
+        
+        print(f"Data for signal '{signal_name}' has been written to {file_path}")
+    except Exception as e:
+        print(f"Error writing data to JSON: {e}")
```
Comment (CMataczynski) [inline Code/utilities.py:72]: warto uważać na nestowane try... except
https://stackoverflow.com/a/49497763

**Review by CMataczynski — APPROVED** [review]

File: Code/signal_cleaning.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
+import numpy as np
+import pandas as pd
+from scipy.interpolate import interp1d
+
+def fill_missing_values(df: pd.DataFrame) -> pd.DataFrame:
```
Comment (auryga) [inline Code/signal_cleaning.py]: any condition in this function? if gap is too long?

File: Code/signal_cleaning.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+            interpolated_values = interpolate_signal(start_index, end_index, 
+                                                    cleaned_signal, method=additional_interpolation_method)
+            cleaned_signal[start_index:end_index] = interpolated_values
+        else:
+            i += 1
+
+    return cleaned_signal
+
+def clean_signal(signal: np.ndarray, window_size: int = 9, threshold_multiplier: float = 2, interpolation_method: str = 'linear',
```
Comment (auryga) [inline Code/signal_cleaning.py:129]: in place where you have interpolation method : by default there we have "linear" - is it possible to change for other?

File: Code/signal_cleaning.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        window = derivative[i-padding:i+padding+1]
+        if np.max(window) >= threshold or np.min(window) <= -threshold:
+            start_index = i - padding
+            end_index = i + padding + 1
+
+            interpolated_values = interpolate_signal(start_index, end_index, 
+                                                     cleaned_signal, method=interpolation_method)
+            cleaned_signal[start_index:end_index] = interpolated_values
+            interpolated_mask[start_index:end_index] = True
+            nan_mask[start_index:end_index] = True  # Here should be logic to set NaN, currently just setting True for the mask
```
Comment (auryga) [inline Code/signal_cleaning.py]: delete such comment from code - if something should be then - should be

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+[tool.poetry]
+name = "sonata-dataset"
+version = "0.1.0"
+description = "This package allows you to experiment with our data smoothly"
+authors = ["Mikolaj Najda <276928@student.pwr.edu.pl>"]
```
Comment (auryga) [inline pyproject.toml:5]: in the PRSA you have gmail here - please decide for one

**Pull request merged**